File: src/AnalyticsService.Database/Repositories/BatchStatRepository.fs
namespace AnalyticsService.Database.Repositories

open AnalyticsService.Database.Model
open AnalyticsService.Database.Model.Domain
open AnalyticsService.Database.Model.Dto
open Dapper.FSharp.PostgreSQL
open System.Threading.Tasks

/// A repository for document batch statistics.
[<Sealed>]
type BatchStatRepository() =
    let statTable = table'<BatchStat> "BatchStat"

    interface IBatchStatRepository with
        member this.AddNewBatchStat stat ctx: Task =
            ctx.Conn.InsertAsync(
                insert {
                    into statTable
                    value stat
                },
                ctx.Tx
            )

        member this.GetBatchStats workflowId conn: Task<BatchStatInfo seq> =
            select {
                for stat in statTable do
                    where (stat.WorkflowId = workflowId)
                    selectAll
            } |> conn.SelectAsync<BatchStatInfo>

        member this.GetBatchStatsForPeriod workflowId startDate period conn: Task<BatchStatInfo seq> =
            let endDate = startDate.Add(period)
            select {
                for stat in statTable do
                    where (stat.WorkflowId = workflowId
                           && stat.DateAdded >= startDate
                           && stat.DateAdded <= endDate)
                    selectAll
            } |> conn.SelectAsync<BatchStatInfo>

File: src/AnalyticsService.Database/Repositories/BatchStatRepository.cs
using System.Data;
using AnalyticsService.Database.Api;
using AnalyticsService.Database.Model.Domain;
using AnalyticsService.Database.Model.Dto;
using AnalyticsService.Database.Queries;
using Dapper;

namespace AnalyticsService.Database.Repositories;

/// <inheritdoc/>
public sealed class BatchStatRepository : IBatchStatRepository
{
    /// <inheritdoc/>
    public async Task AddNewBatchStat(IDbConnection conn, BatchStat stat)
    {
        /*var parameters = new DynamicParameters();
        parameters.Add("@Id", stat.Id);
        parameters.Add("@StartDate", stat.StartDate);*/

        conn.Open();
        using var transaction = conn.BeginTransaction();
        await conn.ExecuteAsync(Query.InsertBatchStat, stat);
        transaction.Commit();
    }

    /// <inheritdoc/>
    public Task<IEnumerable<BatchStatInfo>> GetBatchStats(IDbConnection conn, Guid workflowId)
    {
        conn.Open();
        return conn.QueryAsync<BatchStatInfo>(Query.GetBatchStats, new { WorkflowId = workflowId });
    }

    /// <inheritdoc/>
    public Task<IEnumerable<BatchStatInfo>> GetBatchStats(
        IDbConnection conn,
        DateTime startDate,
        TimeSpan period)
    {
        throw new NotImplementedException();
    }
}